activerecord/lib/active_record/future_result.rb

```diff
@@ -60,15 +66,26 @@ def result
       end
     end
 
-    private
-      def pending?
-        @pending && (!@session || @session.active?)
-      end
+    def instrument(name, payload = {}, &block)
+      @event = @instrumenter.new_event(name, payload)
```

Comment:
I worry that if we receive multiple events it will clobber all but the last. Maybe we could collect them into an array and republish in order?

(next bit non-blocking) It feels like this might be nice to encapsulate in a class. It might be surprising that `instrumenter` is a `FutureResult`, which responds to a lot more than just `instrument`.

Maybe something like... (I'm not attached to the name, `Buffered` already seemed pretty overloaded)

``` ruby
class InstrumentationBuffer
  def initialize(instrumenter)
    @instrumenter = instrumenter
    @events = []
  end

  def instrument(name, payload = {}, &block)
    event = @instrumenter.new_event(name, payload)
    event.record(&block)
    @events << event
  end

  def flush
    @events.each do |event|
      ActiveSupport::Notifications.publish_event(event)
    end
  end
end
```

Reply:
> I worry that if we receive multiple events it will clobber all but the last.

Not sure I follow. We know that a single event happens in that background thread, it's `sql.active_record`, so there's a 1:1 relationship between FutureResult and Event.

> be surprising that instrumenter is a FutureResult, which responds to a lot more than just instrument.

Yeah I expected this feedback :) I'll add an intermediary class, it just that with the two code paths, having FutureResult behave like an instrumenter actually made the code simpler, but yes I kinda agree this is pushing it.

Reply:
Done.